Importing Libraries, dependecies for our program to function:

In [25]:
import numpy as np #we import numpy to work with arrays
import pandas as pd # we import panadas to work and analyse dataframes
from sklearn.model_selection import train_test_split # train_test_split allows us to split our data into training and testing data. Sklearn is a python library from ml data analysis and mining
from sklearn.linear_model import LogisticRegression # we use a logistic regression model hence we import than using the sklearn.linear_model
from sklearn.metrics import accuracy_score # we import accuracy score to evaluate the performance of our learned logistic regression model


In [26]:
# loading our csv file and saving the dataset as pandas dataframe. Panada dataframe is a two dimensional tabular structure, labelled rows and columns, which can be removed as wanted
creditcard_data = pd.read_csv('creditcard_2023.csv')

In [27]:
#printing the first 5 rows of our dataset, note that the time is the time in seconds from the first trascation taking place
creditcard_data.head() #head funtion gives us the first 5 rows, whereas the tail function gives us the last 5 rows

id        V1        V2        V3        V4        V5        V6        V7  \
0   0 -0.260648 -0.469648  2.496266 -0.083724  0.129681  0.732898  0.519014   
1   1  0.985100 -0.356045  0.558056 -0.429654  0.277140  0.428605  0.406466   
2   2 -0.260272 -0.949385  1.728538 -0.457986  0.074062  1.419481  0.743511   
3   3 -0.152152 -0.508959  1.746840 -1.090178  0.249486  1.143312  0.518269   
4   4 -0.206820 -0.165280  1.527053 -0.448293  0.106125  0.530549  0.658849   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0 -0.130006  0.727159  ... -0.110552  0.217606 -0.134794  0.165959  0.126280   
1 -0.133118  0.347452  ... -0.194936 -0.605761  0.079469 -0.577395  0.190090   
2 -0.095576 -0.261297  ... -0.005020  0.702906  0.945045 -1.154666 -0.605564   
3 -0.065130 -0.205698  ... -0.146927 -0.038212 -0.214048 -1.893131  1.003963   
4 -0.212660  1.049921  ... -0.106984  0.729727 -0.161666  0.312561 -0.414116   

        V26       V27       V28    Amount  Class  
0 -0.434824 -0.081230 -0.151045  17982.10    0.0  
1  0.296503 -0.248052 -0.064512   6531.37    0.0  
2 -0.312895 -0.300258 -0.244718   2513.54    0.0  
3 -0.515950 -0.165316  0.048424   5384.44    0.0  
4  1.071126  0.023712  0.419117  14278.97    0.0  

[5 rows x 31 columns]

In our table, we see the amount of each trasncation and then its either classified as a fraudualent transcation or a legit transaction, 0 being legit, 1 being fraudualent

In [28]:
#retreiving credit card information
creditcard_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159688 entries, 0 to 159687
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   id      159688 non-null  int64  
 1   V1      159688 non-null  float64
 2   V2      159688 non-null  float64
 3   V3      159688 non-null  float64
 4   V4      159688 non-null  float64
 5   V5      159688 non-null  float64
 6   V6      159688 non-null  float64
 7   V7      159688 non-null  float64
 8   V8      159688 non-null  float64
 9   V9      159688 non-null  float64
 10  V10     159688 non-null  float64
 11  V11     159688 non-null  float64
 12  V12     159688 non-null  float64
 13  V13     159688 non-null  float64
 14  V14     159687 non-null  float64
 15  V15     159687 non-null  float64
 16  V16     159687 non-null  float64
 17  V17     159687 non-null  float64
 18  V18     159687 non-null  float64
 19  V19     159687 non-null  float64
 20  V20     159687 non-null  float64
 21  V21     15

In [29]:
#since there seems to be some missing values, we confirm if there are any missing values or none by using isnull and sum
creditcard_data.isnull().sum()

id        0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       1
V15       1
V16       1
V17       1
V18       1
V19       1
V20       1
V21       1
V22       1
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [30]:
#column 10-28 has some missing values, we have to balance it. We balance it by using imputation which replaces missing values with substiuted ones
from sklearn.impute import SimpleImputer #this imputes values to balance the missing ones
columns_to_imput = creditcard_data.columns[creditcard_data.isnull().any()].tolist() #selects all columns with missing values for imputation
print(columns_to_imput) # to print columns with missing values
#we will use a mean strategy to impute, hence we create an imputer object with mean strategy
imputer = SimpleImputer(strategy='mean')
creditcard_data[columns_to_imput] = imputer.fit_transform(creditcard_data[columns_to_imput]) # we select the missing columns and impute it using our imputer
#now the columns with missing values are imputed and have the imputer as their substitued value
# this block shows the code used to once impute missing values in our data, hwoever we stored the new data without missing values under a cleaned csv file(later in this code ) and hence dont see any missong values now on
#the cleaned data file is stored later on in this code and hence that file has the modified data from the first time we fixed the missing values

['V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount', 'Class']


In [23]:
#we check and verify that no column has missing values
creditcard_data.isnull().sum()

id        0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [31]:
#none have a missing value as all show columns show a 0 as non-null values.
#now we check for the distribution of our classified data
creditcard_data['Class'].value_counts() # we check for value count of our class column

Class
0.000000    159331
1.000000       356
0.002229         1
Name: count, dtype: int64

In [33]:
#we have a present of 0.002229 value for our class in our data, this is an error in our dataset since a transcation can only be legit (0) oe fraudulent (1).
#Since 0.002229 we convert it to 0. We do this by using replace function
creditcard_data['Class'] = creditcard_data['Class'].replace(0.002229,0)
# Ensure that all values in 'Class' are either 0 or 1
creditcard_data['Class'] = creditcard_data['Class'].apply(lambda x: 1 if x == 1 else 0)
#saving cleaned data
creditcard_data.to_csv('creditcard_cleaned.csv', index=False)


In [34]:
#we check for classified count again to reconfirm
creditcard_data['Class'].value_counts()


Class
0    159332
1       356
Name: count, dtype: int64

In [35]:
#now we balance the data as our two target variables are not balanced. There are 159332 legit transcations and only 356 fraudualent transaction
#We cannot feed this data to our model, hence we balance it
# we seperate the legit and fraudualent dataset for further analysis
legit_data = creditcard_data[creditcard_data.Class == 0]
fraud_data = creditcard_data[creditcard_data.Class == 1]
# we seperate the 0 and 1 column seperately




In [37]:
print(legit_data.shape)
print(fraud_data.shape)

(159332, 31)
(356, 31)


In [38]:
#statistical data about the transcations, so the amount transcated
legit_data.Amount.describe() #gives transcation information about the legit_data
#count is the amount of legit data, mean is the average of all legit transcations, std is the standard deviation from mean and min is minimum transaction value
#the first, second and third quartiles aswell as maximum legit transaction values is stated


count    159332.000000
mean      12033.732454
std        6926.561712
min          50.120000
25%        6041.492500
50%       12001.870000
75%       18039.332500
max       24039.930000
Name: Amount, dtype: float64

In [40]:
fraud_data.Amount.describe() #gives transcation information about the fraud transaction as it did for legit

count      356.000000
mean     11874.509438
std       7085.406101
min         72.820000
25%       5709.035000
50%      11783.765000
75%      17945.362500
max      23987.020000
Name: Amount, dtype: float64

In [41]:
#comparing the two classes of transaction by mean, grouped by mean of every column
creditcard_data.groupby('Class').mean()


id        V1        V2        V3        V4        V5  \
Class                                                                   
0      79848.648784  0.346830 -0.473413  0.957775 -0.665159  0.250917   
1      77539.101124 -0.700801  0.680185 -0.729698  0.834117 -0.593531   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.476289  0.453428 -0.134821  0.608413  ... -0.142739 -0.124528   
1     -0.498900 -0.718174  0.230763 -0.662886  ...  0.196780  0.086133   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.114268 -0.038124  0.145406  0.107717 -0.037699 -0.213492 -0.095876   
1      0.034009  0.013491 -0.039786  0.100217  0.022816  0.275813 -0.000223   

             Amount  
Class                
0      12033.732454  
1      11874.509438  

[2 rows x 30 columns]

In [43]:
#the difference in mean of between fraudualent and legit transcation , will help our model identify legit vs fraud transaction
#balancing data, as there are 159332 legit data and 356 fraudualent data, hence we try balancing it
#we use undersampling to balance the data
#we build a sample data set with similar legit and fraudualent transcartion.
# we get 169 legit and fraud data points
legitsample = legit_data.sample(n=356) #we randomly sample 356 data point from 159332


In [44]:
# we now join/concantenate the two samples together to make a new sample with equal 0 and 1s
new_dataset = pd.concat([legitsample, fraud_data], axis=0)# axis= 0 makes sure that the fraud_data is added below legit sample, row wise and not column wise

In [45]:
# we split the data into test and train set, and train our model with this.
#we check the count of the number of 0 and 1s in our dataset
new_dataset['Class'].value_counts()

Class
0    356
1    356
Name: count, dtype: int64

In [46]:
#checking the new dataset
new_dataset.head()


id        V1        V2        V3        V4        V5        V6  \
132730  132730 -0.246155 -0.489637  0.532579 -0.269172  0.380071  0.606631   
41801    41801  0.948156 -0.428128  0.700883 -0.367412  0.063777  0.001618   
34835    34835  1.035687 -0.064060 -0.169745 -0.160885  1.167354  0.338498   
87193    87193  1.042108 -0.442106  0.543177 -0.476200  0.175642  0.335909   
140009  140009 -0.160261 -0.158307  1.304925 -1.337930  0.093486  0.062723   

              V7        V8        V9  ...       V21       V22       V23  \
132730  1.063765 -0.211905  0.306964  ... -0.118647  0.122314 -0.250024   
41801   0.434589 -0.173711  0.389525  ... -0.053011  0.231935 -0.115430   
34835   0.674235 -0.166986  0.089972  ... -0.187941 -0.451457 -0.392130   
87193   0.371309 -0.127886  0.586622  ... -0.216904 -0.870045  0.016698   
140009  0.467585 -0.014065  0.501826  ... -0.113872 -0.217447  0.003449   

             V24       V25       V26       V27       V28    Amount  Class  
132730 -1.588014 -0.080339 -0.528174  0.463970  0.078578   2591.51      0  
41801   1.268128  0.663662  0.860004 -0.289702 -0.079765  18319.55      0  
34835  -2.550074  1.559540 -0.463796 -0.173146  0.073785  15401.07      0  
87193  -0.942267  0.288297  0.315150 -0.292711 -0.106527   2848.29      0  
140009  0.173758 -0.775603  1.548217 -0.072009  0.148600  15789.21      0  

[5 rows x 31 columns]

In [49]:
new_dataset.shape #checking the shape of our new dataset

(712, 31)

In [51]:
new_dataset.groupby('Class').mean()

id        V1        V2        V3        V4        V5  \
Class                                                                   
0      77687.654494  0.343544 -0.471574  0.947658 -0.687847  0.237572   
1      77539.101124 -0.700801  0.680185 -0.729698  0.834117 -0.593531   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.525482  0.455401 -0.158007  0.604839  ... -0.100126 -0.139297   
1     -0.498900 -0.718174  0.230763 -0.662886  ...  0.196780  0.086133   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.068930 -0.060226  0.056961  0.058123 -0.086803 -0.234323  0.007887   
1      0.034009  0.013491 -0.039786  0.100217  0.022816  0.275813 -0.000223   

             Amount  
Class                
0      12046.651320  
1      11874.509438  

[2 rows x 30 columns]

In [57]:
#splitting the dataset into features and targets, to then split them to test and train

In [59]:
features = new_dataset.drop(columns='Class', axis=1)
targets = new_dataset['Class']
print(features)


            id        V1        V2        V3        V4        V5        V6  \
132730  132730 -0.246155 -0.489637  0.532579 -0.269172  0.380071  0.606631   
41801    41801  0.948156 -0.428128  0.700883 -0.367412  0.063777  0.001618   
34835    34835  1.035687 -0.064060 -0.169745 -0.160885  1.167354  0.338498   
87193    87193  1.042108 -0.442106  0.543177 -0.476200  0.175642  0.335909   
140009  140009 -0.160261 -0.158307  1.304925 -1.337930  0.093486  0.062723   
...        ...       ...       ...       ...       ...       ...       ...   
156990  156990 -0.162200  0.509280 -0.727025  1.354089  0.143071  0.530566   
157585  157585  1.669415 -0.234380  0.354333  0.680745  0.384047  0.418697   
157868  157868 -0.304316  0.218460 -0.777050  0.521586  0.195503 -0.400462   
157871  157871 -0.506172  0.184658 -0.674240 -0.072445 -0.270197  0.035170   
157918  157918 -0.190902 -1.086836  0.249575  1.079678  0.592713  0.721102   

              V7        V8        V9  ...       V20       V21  

In [60]:
print(targets)

132730    0
41801     0
34835     0
87193     0
140009    0
         ..
156990    1
157585    1
157868    1
157871    1
157918    1
Name: Class, Length: 712, dtype: int64


In [61]:
#splitting features and targets into training data and test data
features_train, features_test, targets_train, targets_test = train_test_split(features, targets, test_size=0.15, stratify=targets, random_state=1)
#we split feaTures into test and train, and target into test and train.
# test data size is 15% of overall data and train data size is 85% of overall data, we stratify to targets so that there are equal numbers of 0 and 1 in both sets, random state ensures that the data is split the same way everytime we run this code


In [63]:
print(  features.shape, features_train.shape, features_test.shape)

(712, 30) (605, 30) (107, 30)


In [64]:
print(targets.shape, targets_train.shape, targets_test.shape)


(712,) (605,) (107,)


In [65]:
#Training our logistic regression model
LRmodel = LogisticRegression() #defining our model
LRmodel.fit(features_train, targets_train) #training our model, it fits our model to these data



LogisticRegression()

In [66]:
#evaluating our model performance
#Accuracy on training data
features_train_prediction = LRmodel.predict(features_train)
training_data_accuracy = accuracy_score(targets_train, features_train_prediction)# compring the real target value comprated to what our model predicts on the training data
print('Accuracy on training data : ', training_data_accuracy)

Accuracy on training data :  0.9487603305785124


In [67]:
#Our training data accuracy is 95% which is very good
#Accuracy score on test data, evaluating on test data,which is the main part of our program
features_test_prediction = LRmodel.predict(features_test)
test_data_accuracy = accuracy_score(targets_test, features_test_prediction)
print('Accuracy on test data : ', test_data_accuracy)# we compare what our model predicts for test data to what the actual target value for the test data is

Accuracy on test data :  0.9065420560747663


Our test data accuracy score is 90.65%, which is very good and is not far away from our training data accuracy score. Hence we can say that, our model has not underfitted or overfitted on the training data,  and has good performance and low Generalisation error.